simplipy/api.py

```diff
@@ -259,60 +308,89 @@ def _save_token_data_from_response(self, token_data: dict[str, Any]) -> None:
             self.refresh_token = refresh_token
 
     @staticmethod
-    def is_fatal_error(err: ClientResponseError) -> bool:
+    def is_fatal_error(retriable: list[int]) -> Callable:
         """Determine whether a ClientResponseError is fatal and shouldn't be retried.
 
-        In general, we retry anything outside of HTTP 4xx codes (client errors) with a
-        few exceptions:
+        This call returns the check function.  The arguments to this call are the
+        list of status codes that are retriable.  Generally you'd want to pass
 
         1. 401: We catch this, refresh the access token, and retry the original request.
         2. 409: SimpliSafe base stations regular synchronize themselves with the API,
                 which is where this error can occur; we can't control when/how that
                 happens (e.g., we might query the API in the middle of a base station
                 update), so it should be viewed as retryable.
 
+        But when fetching media files:
+
+        3. 404: When fetching media files, you may get a 404 if the media file is not
+                yet available to read.  Keep trying however, and it will eventually
+                return a 200.
+
         Args:
-            err: An ``aiohttp`` ``ClientResponseError``
+            retriable: A list of retry-able error status codes.
 
         Returns:
-            Whether the error is a fatal one.
+            A callable function used by backoff to check for errors.
         """
-        if err.status in (401, 409):
-            return False
-        return 400 <= err.status < 500
+
+        def check(err: ClientResponseError) -> bool:
```

Comment:
Why are we introducing an inline function here? Can't we just check the statuses and return a bool as before?

Reply:
Its not really an inline function.  check() is returned by is_fatal_error() so it can be used by backoff():

```python
  giveup=self.is_fatal_error(retry_codes)
```

giveup will call a function with (ClientResponseError) and expects a bool in return.  That's check(), but check needs the error codes.  is_fatal_error() supplies the context to the check() function.  I'll make the return type hint on is_fatal_error more explicit if I can.

Reply:
done

Reply:
Take a look at the original method definition: https://github.com/bachya/simplisafe-python/blob/af52ed574ae8d454daa1f5fbcc67ce921c33b0c4/simplipy/api.py#L261-L282

Yours does, indeed, have an inline function. What I'm asking: why can't we just do the check directly within `is_fatal_error` (as before)? You have the list of retry-able error codes available to you there.

Reply:
backoff's "giveup" wants a function that takes an error and expects a boolean.  If is_fatal_error(codes) -> bool is passed like this

```python
    giveup=self.is_fatal_error(codes)
```

then giveup gets a boolean, not a function.  I guess I could do something like this:

```python
    giveup=lambda err: self.is_fatal_error(err, retry_codes)
```

which works, but then I get into mypy lint errors that I don't know how to deal with

    error: Argument 1 to "is_fatal_error" of "API" has incompatible type "Exception"; expected "ClientResponseError"  

If there is another way to do this that I am not aware of, I am totally open to suggestions.

Reply:
This works ... acceptable?

```python
        return backoff.on_exception(
            backoff.expo,
            ClientResponseError,
            giveup=lambda err: self.is_fatal_error(err, retry_codes),  # type: ignore[arg-type] # pylint: disable-line-too-long
            jitter=backoff.random_jitter,
            logger=LOGGER,
            max_tries=request_retries,
            on_backoff=self._async_handle_on_backoff,  # type: ignore[arg-type]
            on_giveup=self._handle_on_giveup,  # type: ignore[arg-type]
        )(request_func)
```

Reply:
Ah, my apologies—I now see what's going on. I've committed some small cleanup, I think your approach is good.

Reply:
Kept current implementation after approval.